WearSpeakerSample/wear/src/main/java/com/example/android/wearable/speaker/SpeakerApp.kt

```diff
@@ -0,0 +1,223 @@
+import androidx.activity.compose.rememberLauncherForActivityResult
+import androidx.activity.result.contract.ActivityResultContracts.RequestPermission
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.LaunchedEffect
+import androidx.compose.runtime.remember
+import androidx.compose.runtime.rememberCoroutineScope
+import androidx.compose.ui.platform.LocalContext
+import androidx.compose.ui.platform.LocalLifecycleOwner
+import androidx.lifecycle.Lifecycle
+import androidx.lifecycle.LifecycleEventObserver
+import androidx.lifecycle.repeatOnLifecycle
+import androidx.wear.compose.material.MaterialTheme
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.CoroutineStart
+import kotlinx.coroutines.async
+import kotlinx.coroutines.coroutineScope
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.selects.select
+import kotlinx.coroutines.suspendCancellableCoroutine
+import kotlin.coroutines.resume
+
+/**
+ * The logic for the speaker sample.
+ *
+ * The stateful logic is kept by a [MainStateHolder].
+ */
+@Composable
+fun SpeakerApp() {
+    MaterialTheme {
+        lateinit var requestPermissionLauncher: ManagedActivityResultLauncher<String, Boolean>
+
+        val context = LocalContext.current
+        val activity = context.findActivity()
+        val scope = rememberCoroutineScope()
+
+        val stateHolder = remember(activity) {
+            MainStateHolder(
+                activity = activity,
+                requestPermission = {
+                    requestPermissionLauncher.launch(Manifest.permission.RECORD_AUDIO)
+                },
+                showPermissionRationale = {
+                    // TODO: Refactor away from normal AlertDialog to a Compose-specific dialog
+                    AlertDialog.Builder(activity)
+                        .setMessage(R.string.rationale_for_microphone_permission)
+                        .setPositiveButton(R.string.ok) { _, _ ->
+                            requestPermissionLauncher.launch(Manifest.permission.RECORD_AUDIO)
+                        }
+                        .setNegativeButton(R.string.cancel) { _, _ -> }
+                        .create()
+                        .show()
+                },
+                showSpeakerNotSupported = {
+                    AlertDialog.Builder(activity)
+                        .setMessage(R.string.no_speaker_supported)
+                        .setPositiveButton(R.string.ok) { _, _ -> }
+                        .create()
+                        .show()
+                }
+            )
+        }
+
+        val lifecycleOwner = LocalLifecycleOwner.current
+
+        requestPermissionLauncher = rememberLauncherForActivityResult(RequestPermission()) {
+            // We ignore the direct result here, since we're going to check anyway.
+            scope.launch {
+                lifecycleOwner.lifecycle.withStateAtLeast(Lifecycle.State.STARTED) {
```

Comment:
Why do you need to wait for the Lifecycle to be at least started to update the state holder?